crypto/x509/x_attrib.c

```diff
@@ -56,3 +56,221 @@ X509_ATTRIBUTE *X509_ATTRIBUTE_create(int nid, int atrtype, void *value)
+        : BIO_printf(out, "%s", objbuf);
+}
+
+int print_attribute_value(BIO *out, int obj_nid, const ASN1_TYPE *av, int indent)
+{
+    const char *ln;
+    char objbuf[80];
+    ASN1_STRING *str;
+    unsigned char *value;
+    X509_NAME *xn = NULL;
+    int64_t int_val;
+
+    // This switch-case is only for syntaxes that are not encoded as a single
+    // primitively-constructed value universal ASN.1 type.
+    switch (obj_nid) {
+    case NID_undef: break; // Unrecognized OID.
+    // Attribute types with DN syntax.
+    case NID_member:
+    case NID_roleOccupant:
+    case NID_seeAlso:
+    case NID_manager:
+    case NID_documentAuthor:
+    case NID_secretary:
+    case NID_associatedName:
+    case NID_dITRedirect:
+    case NID_owner:
+        value = av->value.sequence->data;
+        if ((xn = d2i_X509_NAME(NULL, (const unsigned char**)&(av->value.sequence->data), av->value.sequence->length)) == NULL) {
+            BIO_puts(out, "(COULD NOT DECODE DISTINGUISHED NAME)\n");
+            return 0;
+        }
+        // d2i_ functions increment the ppin pointer. See doc/man3/d2i_X509.pod.
+        // This resets the pointer. We don't want to corrupt this value.
+        av->value.sequence->data = value;
+        if (X509_NAME_print_ex(out, xn, indent, 0) <= 0) {
+            return 0;
+        }
+        X509_NAME_free(xn);
+        return 1;
+
+    default: break;
+    }
+
+    switch (av->type) {
+    case V_ASN1_BOOLEAN:
+        if (av->value.boolean) {
+            return BIO_printf(out, "%*sTRUE", indent, "");
+        } else {
+            return BIO_printf(out, "%*sFALSE", indent, "");
+        }
+
+    case V_ASN1_INTEGER:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        if (ASN1_INTEGER_get_int64(&int_val, av->value.integer) > 0) {
+            return BIO_printf(out, "%ld", int_val);
+        } else {
+            return ASN1_INTEGER_print_bio(out, str);
+        }
+
+    case V_ASN1_ENUMERATED:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        if (ASN1_ENUMERATED_get_int64(&int_val, av->value.enumerated) > 0) {
+            return BIO_printf(out, "%ld", int_val);
+        } else {
+            return ASN1_INTEGER_print_bio(out, str);
+        }
+
+    case V_ASN1_BIT_STRING:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        return print_hex(out, av->value.bit_string->data,
+                 av->value.bit_string->length);
+
+    case V_ASN1_OCTET_STRING:
+    case V_ASN1_VIDEOTEXSTRING:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        return print_hex(out, av->value.octet_string->data,
+                 av->value.octet_string->length);
+
+    case V_ASN1_NULL:
+        return BIO_printf(out, "%*sNULL", indent, "");
+
+    case V_ASN1_OBJECT:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
```

Comment:
shouldn't there be a recursive call here to print the contents of the object?

Reply:
I don't think so. `V_ASN1_OBJECT` is an object identifier, not another recursively structured type. This function also prints attribute values, not ASN.1 values, so recursion doesn't really make sense here unless a particular attribute value is composed of further attributes.

It does however, print structured types in the `V_ASN1_SEQUENCE` and `V_ASN1_SET` cases using the same dump that `asn1dump` uses. But again, recursion doesn't really make sense here, however, because these structure types are not necessarily (and are almost never, if not never) attributes themselves.

I will make this as resolved, but feel free to un-resolve if you have more questions or issues with this.